src/meta/processors/jobMan/JobManager.cpp

```diff
@@ -91,78 +85,210 @@ void JobManager::scheduleThread() {
 
-        // Delete the job after job finished or failed
-        auto it = inFlightJobs_.find(jobDesc->getJobId());
-        if (it != inFlightJobs_.end()) {
-            inFlightJobs_.erase(it);
+        if (!runJobInternal(*jobDesc)) {
+            stopJob(iJob);
         }
     }
     LOG(INFO) << "[JobManager] exit";
 }
 
-// @return: true if succeed, false if any task failed
+// @return: true if all task dispatched, else false
 bool JobManager::runJobInternal(const JobDescription& jobDesc) {
-    currJob_ = MetaJobExecutorFactory::createMetaJobExecutor(jobDesc, kvStore_, adminClient_);
-
-    if (currJob_ == nullptr) {
+    auto jobExec = MetaJobExecutorFactory::createMetaJobExecutor(jobDesc, kvStore_, adminClient_);
+    if (jobExec == nullptr) {
         LOG(ERROR) << "unreconized job cmd " << static_cast<int>(jobDesc.getCmd());
         return false;
     }
+
     if (jobDesc.getStatus() == cpp2::JobStatus::STOPPED) {
-        currJob_->stop();
+        jobExec->stop();
         return true;
     }
 
-    if (!currJob_->check()) {
+    if (!jobExec->check()) {
         LOG(ERROR) << "Job Executor check failed";
         return false;
     }
 
-    if (currJob_->prepare() != cpp2::ErrorCode::SUCCEEDED) {
+    if (jobExec->prepare() != cpp2::ErrorCode::SUCCEEDED) {
         LOG(ERROR) << "Job Executor prepare failed";
         return false;
     }
 
-    auto results = currJob_->execute();
-    if (!nebula::ok(results)) {
-        LOG(ERROR) << "Job executor running failed";
+    if (jobExec->execute() != cpp2::ErrorCode::SUCCEEDED) {
+        LOG(ERROR) << "Job dispatch failed";
         return false;
     }
+    return true;
+}
+
+void JobManager::cleanJob(JobID jobId) {
+    // Delete the job after job finished or failed
+    LOG(INFO) << "[task] cleanJob " << jobId;
+    auto it = inFlightJobs_.find(jobId);
+    if (it != inFlightJobs_.end()) {
+        inFlightJobs_.erase(it);
+    }
+}
+
+cpp2::ErrorCode JobManager::jobFinished(JobID jobId, bool suc) {
+    LOG(INFO) << folly::sformat("{}, jobId={}, suc={}", __func__, jobId, suc);
+    SCOPE_EXIT {
+        cleanJob(jobId);
+        status_ = JbmgrStatus::IDLE;
+    };
+    auto optJobDesc = JobDescription::loadJobDescription(jobId, kvStore_);
+    if (!optJobDesc) {
+        LOG(WARNING) << folly::sformat("can't load job, jobId={}", jobId);
+        return cpp2::ErrorCode::SUCCEEDED;
+    }
+
+    auto jobStatus = suc ? cpp2::JobStatus::FINISHED : cpp2::JobStatus::FAILED;
+
+    optJobDesc->setStatus(jobStatus);
+
+    auto jobExec =
+        MetaJobExecutorFactory::createMetaJobExecutor(*optJobDesc, kvStore_, adminClient_);
+
+    if (jobExec) {
+        if (!optJobDesc->getParas().empty()) {
+            auto spaceName = optJobDesc->getParas().front();
+            auto spaceId = getSpaceId(spaceName);
+            LOG(INFO) << folly::sformat("spaceName={}, spaceId={}", spaceName, spaceId);
+            if (spaceId != -1) {
+                jobExec->setSpaceId(spaceId);
```

Comment:
Maybe we could add a spaceId in thrift `AdminJobReq`, then the paras could just save their own parameters later. And job description could directly use it when job/task added.

Reply:
ok, and I think in another pr is better, in which I can change all the existing job